osu.Game.Tournament/IPC/FileBasedIPC.cs

```diff
@@ -146,54 +156,143 @@ public Storage LocateStableStorage()
+            }
+            finally
+            {
+                Logger.Log($"Stable path for tourney usage: {stableInstallPath}");
+            }
+        }
 
-                    if (checkExists(stableInstallPath))
-                        return stableInstallPath;
-                }
-                catch
-                {
-                }
+        private void saveStablePath()
+        {
+            using (var stream = tournamentStorage.GetStream(stable_config, FileAccess.Write, FileMode.Create))
+            using (var sw = new StreamWriter(stream))
+            {
+                sw.Write(JsonConvert.SerializeObject(stableInfo,
+                    new JsonSerializerSettings
+                    {
+                        Formatting = Formatting.Indented,
+                        NullValueHandling = NullValueHandling.Ignore,
+                        DefaultValueHandling = DefaultValueHandling.Ignore,
+                    }));
+            }
+        }
 
-                stableInstallPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), @"osu!");
-                if (checkExists(stableInstallPath))
-                    return stableInstallPath;
+        private string findFromEnvVar()
+        {
+            try
+            {
+                Logger.Log("Trying to find stable with environment variables");
+                string stableInstallPath = Environment.GetEnvironmentVariable("OSU_STABLE_PATH");
 
-                stableInstallPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".osu");
                 if (checkExists(stableInstallPath))
+                {
+                    stableInfo.StablePath.Value = stableInstallPath;
+                    saveStablePath();
                     return stableInstallPath;
+                }
+            }
+            catch
+            {
+            }
 
-                return null;
+            return null;
+        }
+
+        private string findFromJsonConfig()
+        {
+            try
+            {
+                Logger.Log("Trying to find stable through the json config");
+
+                if (!string.IsNullOrEmpty(stableInfo.StablePath.Value))
+                    return stableInfo.StablePath.Value;
             }
-            finally
+            catch
             {
-                Logger.Log($"Stable path for tourney usage: {stableInstallPath}");
             }
+
+            return null;
+        }
+
+        private string findFromLocalAppData()
+        {
+            Logger.Log("Trying to find stable in %LOCALAPPDATA%");
+            string stableInstallPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), @"osu!");
+
+            if (checkExists(stableInstallPath))
+            {
+                stableInfo.StablePath.Value = stableInstallPath;
+                saveStablePath();
+                return stableInstallPath;
```

Comment:
This scheme of writing to the cached `StableInfo` and then returning the local is very strange. Why not just return `stableInstallPath`, and then assign to `StablePath.Value` and call `saveStablePath` higher up in `findStablePath`?

(Better yet, you could probably assign to the bindable inside of `saveStablePath`, too.)